plugins/experimental/wasm/wasm_main.cc

```diff
@@ -46,6 +46,250 @@ struct WasmInstanceConfig {
+// handler for transform event
+static int
+transform_handler(TSCont contp, ats_wasm::TransformInfo *ti, TSEvent event)
+{
+  TSVConn output_conn;
+  TSVIO input_vio;
+  TSIOBufferReader input_reader;
+  TSIOBufferBlock blk;
+  int64_t toread, towrite, blk_len, upstream_done, input_avail;
+  const char *start;
+  const char *res;
+  size_t res_len;
+  int eos, write_down, empty_input;
+
+  ats_wasm::Context *c;
+
+  TSDebug(WASM_DEBUG_TAG, "[%s] transform handler begins", __FUNCTION__);
+  c = ti->context;
+
+  output_conn = TSTransformOutputVConnGet(contp);
+  input_vio   = TSVConnWriteVIOGet(contp);
+
+  empty_input = 0;
+
+  TSDebug(WASM_DEBUG_TAG, "[%s] cheking input VIO", __FUNCTION__);
+  if (!TSVIOBufferGet(input_vio)) {
+    if (ti->output_vio) {
+      TSDebug(WASM_DEBUG_TAG, "[%s] reenabling output VIO after input VIO does not exist", __FUNCTION__);
+      TSVIONBytesSet(ti->output_vio, ti->total);
+      TSVIOReenable(ti->output_vio);
+      return 0;
+    } else {
+      TSDebug(WASM_DEBUG_TAG, "[%s] no input VIO and output VIO", __FUNCTION__);
+      empty_input = 1;
+    }
+  }
+
+  if (empty_input == 0) {
+    input_reader = TSVIOReaderGet(input_vio);
+  }
+
+  TSDebug(WASM_DEBUG_TAG, "[%s] creating buffer and reader", __FUNCTION__);
+  if (!ti->output_buffer) {
+    ti->output_buffer = TSIOBufferCreate();
+    ti->output_reader = TSIOBufferReaderAlloc(ti->output_buffer);
+
+    ti->reserved_buffer = TSIOBufferCreate();
+    ti->reserved_reader = TSIOBufferReaderAlloc(ti->reserved_buffer);
+
+    if (empty_input == 0) {
+      ti->upstream_bytes = TSVIONBytesGet(input_vio);
+    } else {
+      ti->upstream_bytes = 0;
+    }
+
+    ti->downstream_bytes = INT64_MAX;
+  }
+
+  TSDebug(WASM_DEBUG_TAG, "[%s] init variables inside handler", __FUNCTION__);
+  if (empty_input == 0) {
+    input_avail   = TSIOBufferReaderAvail(input_reader);
+    upstream_done = TSVIONDoneGet(input_vio);
+    toread        = TSVIONTodoGet(input_vio);
+
+    if (toread <= input_avail) { // upstream finished
+      eos = 1;
+    } else {
+      eos = 0;
+    }
+  } else {
+    input_avail   = 0;
+    upstream_done = 0;
+    toread        = 0;
+    eos           = 1;
+  }
+
+  if (input_avail > 0) {
+    // move to the reserved.buffer
+    TSIOBufferCopy(ti->reserved_buffer, input_reader, input_avail, 0);
+
+    // reset input
+    TSIOBufferReaderConsume(input_reader, input_avail);
+    TSVIONDoneSet(input_vio, upstream_done + input_avail);
+  }
+
+  write_down = 0;
+  if (empty_input == 0) {
+    towrite = TSIOBufferReaderAvail(ti->reserved_reader);
+  } else {
+    towrite = 0;
+  }
+
+  do {
+    TSDebug(WASM_DEBUG_TAG, "[%s] inside transform handler loop", __FUNCTION__);
+    proxy_wasm::FilterDataStatus status = proxy_wasm::FilterDataStatus::Continue;
+
+    if (towrite == 0 && empty_input == 0) {
+      break;
+    }
+
+    TSDebug(WASM_DEBUG_TAG, "[%s] retrieving text and calling the wasm handler function", __FUNCTION__);
+    if (empty_input == 0) {
```

Comment:
also this might be more nit-picky but it seems like this conditional is checked a lot and could be refactored to handle this case separately instead of having it checked every block of code